### DISCUSSION TITLE
How to source the setup shell scripts of a ROS 2 workspace?

### DISCUSSION THREAD

**Discussion by valdivj** [iRobotEducation/create3_docs#107] [Questions] [answered]
I am in the create 3 I robot coverage example and I am stuck at a spot I need some help.
The following 2 lines are steps that have to be followed to get example to run.

Build this and the create3_examples_msgs packages. ( I HAVE DONE THIS PART)
Source the setup shell scripts. (HOW DO I DO THIS PART)

**Comment by alsora** [comment, marked as answer]
Hi, the repository `create3_examples` contains various ROS 2 applications to interact with the Create 3 robot.
I recommend you to become familiar with ROS 2 before using it.

"Sourcing the setup shell scripts" is a standard operation that must be done after building any ROS 2 package from sources.
Its purpose is to add the paths to the just built artifacts to the environment paths (e.g. `LD_LIBRARY_PATH`, `PYTHONPATH`, etc).

The command is something like this, depending on the specific paths on your machine.
```
source install/setup.sh
```

Note that full instructions on how to build and source the package can be found in the README here https://github.com/iRobotEducation/create3_examples

If you have issues building/running the `create3_examples` I recommend you to open a Github ticket in that repository.

**Reply by valdivj** [reply]
These are the commands to "Sourcing the setup shell scripts" after you do the build.
You will also have to change the paths to point to the .sh file on you device

For coverage example:
$source /home/joev/create3_examples_ws/build/create3_coverage/ament_cmake_environment_hooks/local_setup.sh

for message example:
$ source /home/joev/create3_examples_ws/build/create3_examples_msgs/ament_cmake_environment_hooks/local_setup.sh

 then this one:
$ source ~/create3_examples_ws/install/local_setup.sh

Then the coverage example should execute the Ros2 commands

**Reply by alsora** [reply]
Hi @valdivj, the commands that you posted are not correct:
 - you should not call the setup scripts inside the `build` directory, but rather those in the `install` directory.
 - you should not call the `local_setup.sh`, but rather the `setup.sh` (unless you have a very specific setup that involves multiple overlays).

Given your specific paths, you only need this command:

```
source ~/create3_examples_ws/install/setup.sh
```

**Reply by valdivj** [reply]
I tried just using 
`source ~/create3_examples_ws/install/setup.sh`
But when I ran
`ros2 action send_goal /coverage create3_examples_msgs/action/Coverage "{explore_duration:{sec: 500, nanosec: 0}, max_runtime:{sec: 1000,nanosec: 0}}"`
it executed the command but the Irobot create 3 did not move

When I ran these 2 scripts after the build of "create3_examples_msgs example" and "create3_coverage example".
`$source /home/joev/create3_examples_ws/build/create3_coverage/ament_cmake_environment_hooks/local_setup.sh`
and 
`$source /home/joev/create3_examples_ws/build/create3_coverage/ament_cmake_environment_hooks/local_setup.sh`

the command:
`ros2 action send_goal /coverage create3_examples_msgs/action/Coverage "{explore_duration:{sec: 500, nanosec: 0}, max_runtime:{sec: 1000,nanosec: 0}}"`
would make the I robot create run.
Honestly I cant nail down why it worked but it was able to work after I did all the above.

**Reply by alsora** [reply]
I can guarantee you that it's just a coincidence.
Sourcing the setup scripts is only needed to let ros know where an executable and its dependencies are.
The way you are doing it is actually dangerous and error-prone.

Sometimes you need to send the first command a couple of times, or start doing a `ros2 topic list` to let the laptop and the robot discover each others

**Reply by valdivj** [reply]
Alsora
  Thanks for the tip.
